zkg

```diff
@@ -284,6 +286,91 @@ def is_git_repo_dirty(git_url):
 
+def regenerate_dependency_state(manager):
+    """
+    Regenerates package dependencies from installed packages.
+    """
+    new_pkgs = []
+
+    to_validate = [(ipkg.package.qualified_name(), ipkg.status.current_version)
+                       for ipkg in manager.installed_packages()]
+
+    invalid_reason, new_pkgs = manager.validate_dependencies(
+            to_validate, ignore_suggestions=True)
+
+    if invalid_reason:
+        print_error('regenerate_dependency_state: failed to resolve dependencies:', invalid_reason)
+        sys.exit(1)
+
+    pkg_dependencies = {}
+
+    for info, version, _ in new_pkgs:
+        name = zeekpkg.package.name_from_path(info.package.git_url)
+        deps = info.dependencies()
+        pkg_dependencies[name] = deps
+
+    return pkg_dependencies
+
+def generate_pkg_dependencies(manager, new_pkgs):
+    pkg_dependencies = load_pkg_dependencies(manager)
+
+    for info, version, _ in new_pkgs:
+        name = zeekpkg.package.name_from_path(info.package.git_url)
+        deps = info.dependencies()
+        pkg_dependencies[name] = deps
+
+    try:
+        dep_pickle_path = os.path.join(manager.scratch_dir, 'deps.pickle')
+        with open(dep_pickle_path, 'wb') as deps:
+           pickle.dump(pkg_dependencies, deps, protocol=pickle.HIGHEST_PROTOCOL)
+    except Exception as exc:
+        zeekpkg.LOG.debug("Error writing dependencies: ", exc)
+
+    manager.set_pkg_dependencies(pkg_dependencies)
+    return pkg_dependencies
+
+def load_pkg_dependencies(manager):
+    try:
+        dep_pickle_path = os.path.join(manager.scratch_dir, 'deps.pickle')
+        with open(dep_pickle_path, 'rb') as deps:
+            return pickle.load(deps)
+    except Exception as exc:
+        zeekpkg.LOG.debug("Error reading dependencies: ", exc)
+        zeekpkg.LOG.debug("Extracting dependencies")        
+        # this is only in case the pickle file is deleted
+        return regenerate_dependency_state(manager)
+
+def loaded_package_states(manager):
+    """Save loaded state for all installed packages.
+
+    Returns:
+        dict: dictionary of load status for installed packages
+    """
+
+    return {name: ipkg.status.is_loaded for name, ipkg in manager.installed_pkgs.items()}
+
+def restoreState(manager, saved_state):
+    """Restores state for installed packages.
+
+    Args:
+        saved_state (dict): dictionary of saved load state for installed
+        packages.
+
+    """
+    try:
+        for _pkg_name in manager.get_pkg_dependencies():
+            ipkg = manager.find_installed_package(_pkg_name)
+            if ipkg and ipkg.status:
+                if ipkg.status.is_loaded == saved_state[_pkg_name]:
+                    continue
+                ipkg.status.is_loaded = saved_state[_pkg_name]
+                manager._write_autoloader()
+                manager._write_manifest()
+                manager._write_plugin_magic(ipkg)
+    except IOError as exc:
+        print(exc)
+        sys.exit(1)
+
```

Comment:
All of these functions seem like they belong in `Manager`, not `zkg`, sorry if I didn't explain well enough before.

The `Manager` object should be able to manage whatever state/persistence upkeep on its own.  Any API call into `Manager` should be assumed that it can work without `zkg` being the thing that is calling it.

Also, the functions that are reading/writing `deps.pickle` don't seem like they'd fit as part of the "public" API.  That is,
`zkg` probably shouldn't have to worry about calling any functions that deal with `deps.pickle`.  I mean, we could document the `Manager` API's to say one needs to do that in order to use these new-and-improved dependency handling features, but it's better to hide that detail from callers if we can.

Reply:
Sorry, I kind of misunderstood your earlier comments. I thought the manager would do just single package management and everything else was to be exposed via zkg. I now get the correct picture and have moved these functions back to the manager code. Also, the functions operating on pickled state are part of manager now.